File: surf-api-bukkit/surf-api-bukkit-api/src/main/java/dev/slne/surf/surfapi/bukkit/api/packet/listener/SurfBukkitPacketListenerApi.java
package dev.slne.surf.surfapi.bukkit.api.packet.listener;

import dev.slne.surf.surfapi.bukkit.api.packet.SurfBukkitPacketApi;
import dev.slne.surf.surfapi.bukkit.api.packet.listener.listener.PacketListener;
import javax.annotation.ParametersAreNonnullByDefault;
import org.jetbrains.annotations.ApiStatus.NonExtendable;

@NonExtendable
@ParametersAreNonnullByDefault
public interface SurfBukkitPacketListenerApi {

  static SurfBukkitPacketListenerApi get() {
    return SurfBukkitPacketApi.get().getPacketListenerApi();
  }

  void registerListeners(final PacketListener listener);

  void unregisterListeners(final PacketListener listener);
}

File: surf-api-bukkit/surf-api-bukkit-api/src/main/kotlin/dev/slne/surf/surfapi/bukkit/api/packet/listener/SurfBukkitPacketListenerApi.kt
package dev.slne.surf.surfapi.bukkit.api.packet.listener

import dev.slne.surf.surfapi.bukkit.api.nms.NmsUseWithCaution
import dev.slne.surf.surfapi.bukkit.api.packet.listener.listener.PacketListener
import dev.slne.surf.surfapi.core.api.util.requiredService

@NmsUseWithCaution
interface SurfBukkitPacketListenerApi {
    fun registerListeners(listener: PacketListener)
    fun unregisterListeners(listener: PacketListener)

    companion object {
        val instance = requiredService<SurfBukkitPacketListenerApi>()
    }
}

@NmsUseWithCaution
val packetListenerApi get() = SurfBukkitPacketListenerApi.instance
